### ISSUE TITLE
Every byte is being sent to the printer twice?

### ISSUE THREAD

**Issue by sh95014** [AppleWin/AppleWin#1017]
Hi! So I thought I'd poke around with the printer support a bit. What I've found is that once I set:

```
    g_bFilterUnprintable = false;
    g_bPrinterAppend = true;
```

I can then tell The Print Shop that we have an "Epson MX-, FX-, RX- (80 or 100)" printer and it would print in AppleWin to `Printer.txt`. Then, with <100 lines (of very ugly code that's not worth sharing yet) I can get the bits out:

![printshop](https://user-images.githubusercontent.com/95387068/148394269-f97cbd4d-d5b7-4dac-a319-89708885436e.png)

so the good news is that I don't think we'll need a lot of printer command implementation to get pretty good support.

**My big question**, however, is why the output captured in `Printer.txt` appears to send every byte twice, as my hex editor shows in the screenshot below. My ugly code just deliberately drops every other byte, but I was wondering if anybody already knows why. I don't have a PC so I can't see what AppleWin is doing under Windows and I was hoping somebody would try the flags I set above and see whether they also get the doubling from The Print Shop. I believe `PrintTransmit()` is actually getting called twice for each byte. Simply doing a `PR#1` doesn't seem to do that!

**My other question** is why `g_bFilterUnprintable` defaults to true, and why that's an option at all. Seems like any filtering should be done on the "printer" side and not the controller side, but I imagine I'm missing something.

Cheers!

<img width="555" alt="Screen Shot 2022-01-06 at 10 17 38 PM" src="https://user-images.githubusercontent.com/95387068/148396830-260d8698-20dc-4257-9112-c17543139a63.png">
...

Gory Details:

Basically the bitmap is sent in a simple command like this:

ESC  'L'  length-lo  length-hi  [length bytes]

and the printer just sends a LF (and CR) to move the print head when needed. It also uses another `<ESC> 'A'` command to set the line height (how much LF moves the paper vertically) so that there won't be a gap in between passes.

Within the data bytes, the print head is 9 pins arranged vertically (so it can print a line of text in one pass) but I think the Apple parallel card can only address 7 of those pins, so the MSB of the byte is unused.

So basically I have a buffer of 7 rows of pixels, and the 7 meaningful bits of each byte from the printer is distributed to the 7 rows, one bit at a time.

Once I'm done with the row (i.e., received a LF) I just printed out the bits in XBM format and got it converted to PNG on some website.

**Comment by tomcw** [comment]
Hi Nick / @sicklittlemonkey - this reminds me of your "Ancient Printer Emulator" patch that you emailed some time ago (2017 in fact!). As a minimum I think I'll create a new issue and attach your patch zip so that it is closer to the AppleWin action!

**Comment by tomcw** [comment]
Hi @sh95014 - this looks good, thanks for sharing!

I also quickly tried `PR#1` and some simple typing etc. and it only outputs a single character.

>My big question, however, is why the output captured in Printer.txt appears to send every byte twice,

I don't know. I'd guess that because you select the "Epson MX-, FX-, RX- (80 or 100)" printer, then the firmware (in AppleWin's slot 1 printer card) isn't honouring some command correctly? Do you also select a printer card?

>My other question is why g_bFilterUnprintable defaults to true, and why that's an option at all.

This was added by someone else who also added the Pravets clone support. I imagine that lots of non ASCII characters were being output to the printer.txt file, and this was causing problems.

So your suggestion is just remove g_bFilterUnprintable, or default it to false (and have some option to enable it).

**Comment by sh95014** [comment]
> I don't know. I'd guess that because you select the "Epson MX-, FX-, RX- (80 or 100)" printer, then the firmware (in AppleWin's slot 1 printer card) isn't honouring some command correctly? Do you also select a printer card?

Yes, after the Epson, I selected "Apple II Parallel Interface" in slot 1, with 2 disk drives.

> > My other question is why g_bFilterUnprintable defaults to true, and why that's an option at all.
> 
> This was added by someone else who also added the Pravets clone support. I imagine that lots of non ASCII characters were being output to the printer.txt file, and this was causing problems.
> 
> So your suggestion is just remove g_bFilterUnprintable, or default it to false (and have some option to enable it).

Yeah, I think so. It's filtering out the ESC which is the prefix character for all the Epson commands.

**Comment by sh95014** [comment]
@tomcw Had a bit of time to look into this, and I know what's going on now. PrintShop talks to the Parallel Interface Card (PIC) directly, so it writes the character, sends a strobe, and polls for an acknowledgement. Adding a simple log to `PrintStatus()` and `PrintTransmit()` shows this pattern of access:

```
PRINTER WRITE $1BF1 $C090 1:$0D 62
PRINTER WRITE $1BF4 $C092 1:$0D 67
PRINTER READ $1BE9 $C094 0:$00 17
```

In this case, it writes $0D to $C090 (output port), then the same value to $C092 (strobe), which I believe is a signal to the printer to pick up the value latched at $C090. It then reads $C094 (acknowledge) and is presumably happy with the hardcoded $FF it gets back from `PrintStatus()`.

So the bug here is that `PrintTransmit()` does not correctly handle writes to $C092 and thinks it's another data byte being sent to the printer. The fix would be pretty straightforward.

My question is broader, however. I'd like to implement the graphics printing support for maybe a MX-80, but I want to put the code in the right place. In my mind, `ParallelPrinter.cpp` should basically emulate the PIC but not the printer, and instead of writing its output to a file, it should send it to a new `MX80.cpp` which might save it to text files or BMP files. But this would be a far more substantial refactoring of `ParallelPrinter.cpp` so I'd like to know if that makes sense to you and the other maintainers, but it just seems tedious to me if the printer emulation has to go through the disk file that `ParallelPrinter.cpp` saves and can unexpectedly wipe after some period of inactivity. Using that as API just seems hopelessly brittle. A proper split would also allow other printers to be emulated later.

The smaller question is the behavior of the "paper." I'm thinking:
- We start a new text file every time the emulation is rebooted, as if you turn the printer off and on along with the emulated computer.
- Graphics commands/data are stripped from the text file, and instead a BMP file is created for each page (separated by the Form Feed character).
- Text and BMP files older than some (configurable?) interval (say, 30 days by default) can be deleted, but that doesn't have to be implemented in the printer code at all.

Sorry it's a wall of text, but does all that make sense? If not I'll just make the changes downstream instead.

Thanks!

**Comment by tomcw** [comment]
Hi @sh95014,

>So the bug here is that `PrintTransmit()` does not correctly handle writes to $C092 and thinks it's another data byte being sent to the printer. The fix would be pretty straightforward.

OK, so let's fix that first. Perhaps you can submit a PR for this.

>I'd like to implement the graphics printing support for maybe a MX-80, but I want to put the code in the right place.

Great! Please start by reading this: https://github.com/AppleWin/AppleWin/blob/master/CONTRIBUTING.md

I recommend initially adding support via the command line, and not trying to shoehorn extra GUI controls into the Configuration prop-sheet-pages.

There was also another PoC done by Nick (@sicklittlemonkey) a few years ago. I will attached it here for reference.

**Comment by tomcw** [comment]
OK, I've added Nick's PoC details & code drop (in a zip) in #1026.

**Comment by sicklittlemonkey** [comment]
Thanks Tom. This actually worked pretty well apart from the graphics fudge
factor problem.
Given the cross-platform nature of AppleWin now, I wonder how
GdiPrinterOutput would translate to Linux.
In any case it really needs a PdfFileWriter using a PDF library. That might
fix both problems.

**Comment by sh95014** [comment]
@tomcw I can't seem to open a PR. Some sort of permission issue?

Anyway, here's the details and the diff:

> Only send writes to address $C090 (although the code is generalized in case the PCI is inserted into a different slot) to the printer. PrintShop was also writing to the strobe ($C092) which should not be interpreted as a data byte.
> 
> Tested with PrintShop configured for a "Epson MX-, FX-, RX- (80 or 100)" connected to "Apple II Parallel Interface" in slot 1.
> 
> This still does *not* cleanly pass the bytes to an Epson printer. You'll also want to set g_bFilterUnprintable to false because that filters out the ESC character that the printer needs, and also g_bPrinterAppend because that has a bad habit of resetting the printout.

```
diff --git a/source/ParallelPrinter.cpp b/source/ParallelPrinter.cpp
index 38fff52c..5c05d25d 100644
--- a/source/ParallelPrinter.cpp
+++ b/source/ParallelPrinter.cpp
@@ -142,21 +142,25 @@ static BYTE __stdcall PrintStatus(WORD, WORD, BYTE, BYTE, ULONG)
 }
 
 //===========================================================================
-static BYTE __stdcall PrintTransmit(WORD, WORD, BYTE, BYTE value, ULONG)
+static BYTE __stdcall PrintTransmit(WORD, WORD address, BYTE, BYTE value, ULONG)
 {
        if (!CheckPrint())
                return 0;
 
-       BYTE c = value & 0x7F;
+    if ((address & 0x0F) == 0)
+    {
+        // load output port (i.e., $C090)
+       BYTE c = value & 0x7F;
 
-       if (IsPravets(GetApple2Type()))
-       {
-               if (g_bConvertEncoding)
-                       c = GetPravets().ConvertToPrinterChar(value);
-       }
+       if (IsPravets(GetApple2Type()))
+       {
+               if (g_bConvertEncoding)
+                       c = GetPravets().ConvertToPrinterChar(value);
+       }
 
-       if ((g_bFilterUnprintable == false) || (c>31) || (c==13) || (c==10) || (c>0x7F)) //c>0x7F is needed for cyrillic characters
-               fwrite(&c, 1, 1, file);
+       if ((g_bFilterUnprintable == false) || (c>31) || (c==13) || (c==10) || (c>0x7F)) //c>0x7F is needed for cyrillic characters
+               fwrite(&c, 1, 1, file);
+    }
```

**Comment by tomcw** [comment]
You'll need to fork AppleWin to your own GitHub user space, branch it, make this change... and only then you'll be able to submit a PR.

Can you try to get this working, as you'll need to be able to do this for the bigger MX80 piece of work.

Thanks.

**Comment by sh95014** [comment]
@tomcw I suck at git/GitHub, but here you go: https://github.com/AppleWin/AppleWin/pull/1027

**Comment by tomcw** [comment]
Closing this issue, as the original problem (every byte sent to printer twice) was fixed in PR #1027, and will be in AppleWin 1.30.9.0 (date still TBD). And more recent and active discussion around Nick's "APEL" based implementation is in #1026.

**Comment by tomcw** [comment]
FYI: new AppleWin 1.30.9.0 release [here](https://github.com/AppleWin/AppleWin/releases/tag/v1.30.9.0) with this fixed.
